[NJT] Support NJT SDPA + meta-device flop counting

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #134289
* #134288

A user wants to use the flop counter with meta devices. This previously caused problems for SDPA+NJT:

1. autocast check: `torch.is_autocast_enabled("meta")` fails because `meta` is not valid for autocasting. If we skip this, we run into the next error
2. math backend: conversion to NST requires getting concrete offsets in a list of python integers, which doesn't work on a meta tensor https://github.com/pytorch/pytorch/blob/b2eb0e8c6a817f304277dfed39c25853ff301d90/torch/nested/_internal/sdpa.py#L809-L815
3. (fixed in the previous PR, #134288) - if we force using flash attention backend for flop counting, `_flash_attention_forward` previously didn't support meta tensors.

In this PR, we check specifically for FlopCounterMode, and, if it's enabled and combined with meta tensors, (a) skip autocasting and (b) force it down the flash attention path. This isn't generally safe for tracing (e.g. if you actually care which kernels you are running), but in the absence of actual device information, we have to make some assumptions. By specifically checking for FlopCounterMode, this should reduce the chance of unintended side effects for other meta tensor users.

Note: fake tensor would solve a bunch of these issues, but it's not a viable solution right now for the user.

cc @cpuhrsch @jbschlosser @bhosmer @drisspg @soulitzer @YuqingJ

davidberard98 committed:
[NJT] Support NJT SDPA + meta-device flop counting

[ghstack-poisoned]

davidberard98 committed:
Update on "[NJT] Support NJT SDPA + meta-device flop counting"




[ghstack-poisoned]

soulitzer commented:
I think you mentioned that you were looking into just getting FakeTensors to work, curious what issues you ran into there. 

Moving the logic meta registrations sounds good anyway though, like I'd imagine shape propagation stuff to tend to belong in meta registrations, and then you'd wrap things in the correct devices in the Fake layer.

Re @YuqingJ's approach to wrap NST logic in an operators, I guess that requires us to make math no longer composite, and then to handle things like autograd/vmap?

soulitzer commented on torch/nested/_internal/sdpa.py:
```diff
@@ -642,7 +656,8 @@ def _autocast(
     device_type = query.device.type
-    if not torch.is_autocast_enabled(device_type):
+    # meta device is not supported by autocast, so break early for it
+    if _is_computing_meta_flops(query) or not torch.is_autocast_enabled(device_type):
```
Does this not fail usually anyway?

davidberard98 replied:
`torch.is_autocast_enabled("meta")` throws an error

davidberard98 commented:
@soulitzer 

> Re @YuqingJ's approach to wrap NST logic in an operators, I guess that requires us to make math no longer composite, and then to handle things like autograd/vmap?

another possibility is that we wrap the NJT -> NST in an op. But IDK how meta tensors would work with NST - like what size would you give the meta tensor representing the NST?

davidberard98 committed:
Update on "[NJT] Support NJT SDPA + meta-device flop counting"


A user wants to use the flop counter with meta devices. This previously caused problems for SDPA+NJT:

1. autocast check: `torch.is_autocast_enabled("meta")` fails because `meta` is not valid for autocasting. If we skip this, we run into the next error
2. math backend: conversion to NST requires getting concrete offsets in a list of python integers, which doesn't work on a meta tensor https://github.com/pytorch/pytorch/blob/b2eb0e8c6a817f304277dfed39c25853ff301d90/torch/nested/_internal/sdpa.py#L809-L815
3. (fixed in the previous PR, #134288) - if we force using flash attention backend for flop counting, `_flash_attention_forward` previously didn't support meta tensors.

In this PR, we check specifically for FlopCounterMode, and, if it's enabled and combined with meta tensors, (a) skip autocasting and (b) force it down the flash attention path. This isn't generally safe for tracing (e.g. if you actually care which kernels you are running), but in the absence of actual device information, we have to make some assumptions. By specifically checking for FlopCounterMode, this should reduce the chance of unintended side effects for other meta tensor users.

Note: fake tensor would solve a bunch of these issues, but it's not a viable solution right now for the user.

cc cpuhrsch jbschlosser bhosmer drisspg soulitzer YuqingJ

[ghstack-poisoned]

davidberard98 committed:
Update on "[NJT] Support NJT SDPA + meta-device flop counting"


A user wants to use the flop counter with meta devices. This previously caused problems for SDPA+NJT:

1. autocast check: `torch.is_autocast_enabled("meta")` fails because `meta` is not valid for autocasting. If we skip this, we run into the next error
2. math backend: conversion to NST requires getting concrete offsets in a list of python integers, which doesn't work on a meta tensor https://github.com/pytorch/pytorch/blob/b2eb0e8c6a817f304277dfed39c25853ff301d90/torch/nested/_internal/sdpa.py#L809-L815
3. (fixed in the previous PR, #134288) - if we force using flash attention backend for flop counting, `_flash_attention_forward` previously didn't support meta tensors.

In this PR, we check specifically for FlopCounterMode, and, if it's enabled and combined with meta tensors, (a) skip autocasting and (b) force it down the flash attention path. This isn't generally safe for tracing (e.g. if you actually care which kernels you are running), but in the absence of actual device information, we have to make some assumptions. By specifically checking for FlopCounterMode, this should reduce the chance of unintended side effects for other meta tensor users.

Note: fake tensor would solve a bunch of these issues, but it's not a viable solution right now for the user.

cc cpuhrsch jbschlosser bhosmer drisspg soulitzer YuqingJ

[ghstack-poisoned]

davidberard98 committed:
Update on "[NJT] Support NJT SDPA + meta-device flop counting"


A user wants to use the flop counter with meta devices. This previously caused problems for SDPA+NJT:

1. autocast check: `torch.is_autocast_enabled("meta")` fails because `meta` is not valid for autocasting. If we skip this, we run into the next error
2. math backend: conversion to NST requires getting concrete offsets in a list of python integers, which doesn't work on a meta tensor https://github.com/pytorch/pytorch/blob/b2eb0e8c6a817f304277dfed39c25853ff301d90/torch/nested/_internal/sdpa.py#L809-L815
3. (fixed in the previous PR, #134288) - if we force using flash attention backend for flop counting, `_flash_attention_forward` previously didn't support meta tensors.

In this PR, we check specifically for FlopCounterMode, and, if it's enabled and combined with meta tensors, (a) skip autocasting and (b) force it down the flash attention path. This isn't generally safe for tracing (e.g. if you actually care which kernels you are running), but in the absence of actual device information, we have to make some assumptions. By specifically checking for FlopCounterMode, this should reduce the chance of unintended side effects for other meta tensor users.

Note: fake tensor would solve a bunch of these issues, but it's not a viable solution right now for the user.

cc cpuhrsch jbschlosser bhosmer drisspg soulitzer YuqingJ

[ghstack-poisoned]

soulitzer reviewed (approved):
SGTM

davidberard98 commented:
@pytorchbot merge
